= BSB - Databases
:toc:
:toc-title: Contents
:nofooter:
:stem: latexmath

== Where do we find information?

There are a whole bunch of interlinked databases

. Nucleotide sequences:
* ENA
* GenBank
* INDSC
. Genes
* NCBI Gene
** proteins and the genes that produce them have the same name (lol???) but different IDs
** different research groups can name the same protein/gene/whatever differently
*** databases list synonyms as well
* GeneCards
** less about data
** more about generic information
. Genomes/proteome/transcriptome
* GenBank
. Genome Browsers
* ensembl
** shows gene location in genome
** shows gene isoforms
** autologs: same gene different life form
** paralogs: something else idk
** shows phenotypes: expression of genes (eye color, hair color, pea color, pea texture, etc.)

Most of these have APIs to make automatic interaction easier

== How is this information written?

=== FASTA

Seen before

Used for known sequences

ATAGCAC and such

FASTA has a sibling type called fastq which is formatted like this:

[source,python]
----
@FORJUSP02AJWD1
CCGTCAATTCATTTAAGTTTT
+
AAAAAAAAAAAA::99@::::
----

* the first line is a sequence label
* the second is the raw sequence
* the `+` sign to act as separator
* quality values for each element of the raw sequence

The quality values correspond to ASCII values of their character


CAUTION: the first 33 ASCII characters are unprintable. Offset the ASCII code by 33 to get the correct score

WARNING: offset older (no explanation as to what older means) by 64 instead


[source,python]
----
q = "AAAAAAAAAAAA::99@::::"

q_converted = " ".join(ord(c) - 33 for c in q)
----

For each character in the quality string we can compute the *Base Quality score* (or Phred score)

* indicates the probability that a given base is incorrectly called by the sequencer
** sequencer says `C` is at index `10` when it's really `A`
** what are the odds of that?

==== BQ/Phred score

Higher is better

.Formula (that gets you `Q`)
[stem]
++++
Q = -10 \log_{10}(e)
++++

Where:

* stem:[e] is the estimated probability of that base call being wrong
** not sure why they used stem:[e] instead of stem:[p] to denote probability but whatever

If you have `Q` (which you do from the FASTQ file) you use this to get Phred:

Lower is better

[stem]
++++
P = 10^{\frac{-Q} {10}}
++++

and you could use something like this to get the probability:

[source,python]
----
def phred(q: str) -> float:
    q = ord(q) - 33
    return pow(10, (-q / 10))

p = phred("A") # ~0.0006
----

There are 3 main quality levels (but there are more):

* stem:[10 \rightarrow (Q10)]
** 1 in 10 chance of base call being wrong
** 90% inferred base call accuracy

* stem:[20 \rightarrow (Q20)]
** 1 in in 100 chance of base call being wrong
** 99% inferred base call accuracy

* stem:[30 \rightarrow (Q30)]
** 1 in 1000 chance of base call being wrong
** 99.9% inferred base call accuracy

=== Transcript Annotation Files

Used to encode transcripts

. https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md[GFF]
* GFF2 represented 2 levels of feature hierarchy
* GFF3 can represent any number of features in a hierarchy
. GTF
* identical to GFF2

GFF files:

* Plain text 
* 9 columns
** code
** gene part
** start point
** end point
** other stuff
* tab delimited

=== UTR

Regions on RNA extremes remain untranslated

They are required for translation but they themselves are not translated

* key regulatory regions

One at 3'

* regulates gene expression
* influences mRNA stability
* affects within-cell localization
* binds to regulatory molecules
** microRNAs
** other RNA binding proteins

One at 5'

* located before the start codon (AUG)
* regulates translation
** controls ribosome binding
** mRNA stability
** efficiency of protein synthesis

== Sequence Analysis

How do we analyze these sequences?

=== Pattern Matching

Similar patterns can indicate similar functions

Can be used to reconstruct genome/gene/protein evolution

2 main ways:

* exact pattern matching
** find ATAGC in human genome
* inexact pattern matching
** A*AGC in human genome
** nightmare in bioinf

Need the right algorithm for the job

.Knuth-Morris-Pratt (KMP) algorithm
. preprocess steps
* construct Longest Proper Prefix that is also a Suffix
** when a mismatch occurs the word has enough information to guess where the next match could begin

.Regular Expressions
* Useful in genomes because fixed sequences
* Letters outside ATGCU signify other things
** nucleotide IUPAC
* Amino acids only use IUPAC
** no weirdness

=== Codons to amino acids

3 codons make an amino acid

Wheel that maps starting codon to amino acids

* U -> U -> U or C => phenyline

WARNING: this wheel does not apply to all living beings. Some bacteria and human mithocondria have exceptions

== Homework

=== HW1: TP53

==== Humans
* https://www.ncbi.nlm.nih.gov/nuccore/NM_000546.6?report=fasta[Genomic sequence]
* https://www.ncbi.nlm.nih.gov/datasets/gene/7157/#transcripts-and-proteins[Derived protein sequences]
* https://www.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000141510;r=17:7661779-7687546;t=ENST00000714409[Longest transcript]
** 39 splice variants
** 34 protein coding
** https://www.ensembl.org/Homo_sapiens/Transcript/Sequence_cDNA?db=core;g=ENSG00000141510;r=17:7661779-7687546;t=ENST00000714409[Coding sequence]
* https://sep2025.archive.ensembl.org/Homo_sapiens/Gene/Compara_Tree?db=core;g=ENSG00000141510;r=17:7661779-7687546[Evolutionary tree]

==== Mice
* https://www.ncbi.nlm.nih.gov/nuccore/NM_011640.4?report=fasta[Genomic sequence]
* https://www.ncbi.nlm.nih.gov/datasets/gene/22059/#transcripts-and-proteins[Derived protein sequences]

==== Chimps
* https://www.ensembl.org/Pan_troglodytes/Gene/Sequence?g=ENSPTRG00000008703;r=17:7936163-7951161[Genomic sequence]
* https://www.ncbi.nlm.nih.gov/datasets/gene/455214/#transcripts-and-proteins[Derived protein sequences]
